graphql_compiler/tests/snapshot_tests/test_cost_estimation.py

```diff
@@ -131,6 +131,163 @@ def test_complex_traverse(self):
         )
         self.assertAlmostEqual(expected_cardinality_estimate, cardinality_estimate)
 
+    @pytest.mark.usefixtures('snapshot_orientdb_client')
+    def test_traversal_provided_both_statistics(self):
+        """Test type coercion provided both class_counts and vertex_edge_vertex_counts."""
+        schema_graph = generate_schema_graph(self.orientdb_client)
+        graphql_input = '''{
+            Animal {
+                out_Entity_Related {
+                    ... on Event {
+                        uuid @output(out_name: "event_id")
+                    }
+                }
+            }
+        }'''
+        params = {}
+
+        count_data = {
+            'Entity': 19,
+            'Animal': 3,
+            'Event': 7,
+            'Entity_Related': 11
+        }
+        vertex_edge_vertex_data = {
+            ('Animal', 'Entity_Related', 'Event'): 2
+        }
+        statistics = LocalStatistics(
+            count_data, vertex_edge_vertex_counts=vertex_edge_vertex_data
+        )
+
+        cardinality_estimate = estimate_query_result_cardinality(
+            schema_graph, statistics, graphql_input, params
+        )
+
+        # For each Animal, vertex_edge_vertex counts tell us we should expect (2.0 / 3.0) Events
+        # when traversing using Entity_Related. This totals to 3.0 * (2.0 / 3.0) results.
+        expected_cardinality_estimate = 3.0 * (2.0 / 3.0)
+        self.assertAlmostEqual(expected_cardinality_estimate, cardinality_estimate)
+
+    @pytest.mark.usefixtures('snapshot_orientdb_client')
+    def test_traversal_with_no_results(self):
+        """Test type coercion where no results should be expected."""
+        schema_graph = generate_schema_graph(self.orientdb_client)
+        graphql_input = '''{
+            Animal {
+                out_Entity_Related {
+                    ... on Event {
+                        uuid @output(out_name: "event_id")
+                    }
+                }
+            }
+        }'''
+        params = {}
+
+        count_data = {
+            'Entity': 19,
+            'Animal': 3,
+            'Event': 7,
+            'Entity_Related': 11
+        }
+        vertex_edge_vertex_data = {
+            ('Animal', 'Entity_Related', 'Event'): 0
+        }
+        statistics = LocalStatistics(
+            count_data, vertex_edge_vertex_counts=vertex_edge_vertex_data
+        )
+
+        cardinality_estimate = estimate_query_result_cardinality(
+            schema_graph, statistics, graphql_input, params
+        )
+
+        # Vertex_edge_vertex_data tells us that no Entity_Related edges connect Animals and Events,
+        # so the result set is empty.
+        expected_cardinality_estimate = 0.0
+        self.assertAlmostEqual(expected_cardinality_estimate, cardinality_estimate)
+
+    @pytest.mark.usefixtures('snapshot_orientdb_client')
+    def test_traversal_in_reverse_direction_provided_both_statistics(self):
```

Comment:
nit: inbound instead of reverse to consistent with the docstring.

Reply:
Good point!